python/cac_tripplanner/destinations/views.py

```diff
@@ -378,16 +379,87 @@ def get(self, request, *args, **kwargs):
             try:
                 limit_int = int(limit)
             except ValueError as e:
-                error = json.dumps({
-                    'msg': 'Invalid limit, must be an integer',
-                    'error': str(e),
-                })
-                return HttpResponse(error, 'application/json')
+                return return_400('Invalid limit, must be an integer', str(e))
             destinations = destinations[:limit_int]
             events = events[:limit_int]
 
         destinations = [set_destination_properties(x) for x in destinations]
         events = [set_event_properties(x) for x in events]
 
         response = {'destinations': destinations, 'events': events}
-        return HttpResponse(json.dumps(response), 'application/json')
+        return JsonResponse(response)
+
+
+def return_400(message, error):
+    # Helper to return JSON error messages in a consistent format
+    error = {
+        'msg': message,
+        'error': error
+    }
+    return JsonResponse(error, status=400)
+
+
+class UserFlagView(View):
+    """POST-only endpoint for recording anonymous user flags on destinations from mobile app.
+
+    Expects POST as JSON with the following fields:
+     - api_key: String key that must match setting from server secrets file
+     - attraction: ID of the destination or event to flag
+     - is_event: true if attraction is an event
+     - user_uuid: UUID to anonymously identify user (not associated with anything else)
+     - flag: how user has flagged the attraction; must be one of `UserFlag.UserFlags` keys
+             (or empty, if unset)
+    """
+    @method_decorator(csrf_exempt)
+    def dispatch(self, request, *args, **kwargs):
+        return super(UserFlagView, self).dispatch(request, *args, **kwargs)
+
+    def post(self, request, *args, **kwargs):
+        # decode posted body for python 3 support
+        try:
+            json_data = json.loads(request.body.decode('utf-8'))
+        except ValueError as e:
+            return return_400('Could not parse JSON', str(e))
+
+        if json_data.get('api_key') != settings.USER_FLAG_API_KEY:
+            return return_400('api_key required', 'API key missing or incorrect')
+
+        expected_fields = ('attraction', 'is_event', 'user_uuid', 'flag')
+        user_flag_data = {}  # build an object of the cleaned values to use to create flag
+        missing_fields_msg = 'Missing expected value(s): '
+        missing = False  # flag to check if any of the required fields are missing
+        for fld in expected_fields:
+            if fld in json_data:
+                user_flag_data[fld] = json_data.get(fld, None)
+            else:
+                missing = True
+                missing_fields_msg += fld + ', '
+        if missing:
+            return return_400(missing_fields_msg, 'Missing required value')
+
+        # parse string JSON value to boolean
+        user_flag_data['is_event'] = user_flag_data['is_event'] == 'true'
```

Comment:
The `is_event` property is already coming through as a boolean value, not a string, so this line needs to be removed or all flags are interpreted as Destinations.